coordinator/src/main/scala/com/socrata/datacoordinator/service/Main.scala

```diff
@@ -116,46 +119,86 @@ class Main(common: SoQLCommon, serviceConfig: ServiceConfig) {
+
+            if(schemaHash.isDefined && (Some(schema.hash) != schemaHash)) {
+              f(Left(schema))
+            } else {
+              val jsonReps = common.jsonReps(copyCtx.datasetInfo)
+              val jsonSchema = copyCtx.schema.mapValuesStrict { ci => jsonReps(ci.typ) }
+              val unwrappedCids = copyCtx.schema.values.toSeq.filter { ci => jsonSchema.contains(ci.systemId) }
+                .sortBy(_.userColumnId).map(_.systemId.underlying).toArray
+              val pkColName = copyCtx.pkCol.map(_.userColumnId)
+              val orderedSchema = unwrappedCids.map { cidRaw =>
+                val col = copyCtx.schema(new ColumnId(cidRaw))
+                SchemaField(col.userColumnId, col.typ.name.name)
               }
-              arr
-            }))
-          )
+              f(Right((
+                entityTag,
+                orderedSchema,
+                pkColName,
+                copyCtx.datasetInfo.localeName,
+                approxRowCount,
+                it.map { row =>
+                  val arr = new Array[JValue](unwrappedCids.length)
+                  var i = 0
+                  while(i != unwrappedCids.length) {
+                    val cid = new ColumnId(unwrappedCids(i))
+                    val rep = jsonSchema(cid)
+                    arr(i) = rep.toJValue(row(cid))
+                    i += 1
+                  }
+                  arr
+                }))
+              )
+            }
+          }
+        case Left(err) => err
+      }
+    }
+  }
+
+  /**
+   * Read row identifier value in string form into SoQLValue.
+   * It is an error only when the column rep cannot parse a non-empty row id value.
+   * It is not an error if row id value does not exist but is valid.
+   */
+  private def readRowId(u: PostgresUniverse[SoQLType, SoQLValue],
+                        id: DatasetId,
+                        copy: CopySelector,
+                        rowId: Option[String]): Either[Exporter.Result[Unit], Option[SoQLValue]] = {
+    val rowIdPossiblyError =
+      for {
+        rid <- rowId
+        ctxOpt <- u.datasetReader.openDataset(id, copy)
+        ctx <- ctxOpt
+      } yield {
+        ctx.copyCtx.userIdCol match {
+          case Some(_) => // dataset has custom row identifier
+            val rowIdRep = SoQLRep.csvRep(ctx.copyCtx.pkCol_!)
+            // optional row id type must be simple which is represented by a single csv string.
+            rowIdRep.decode(IndexedSeq(rid), IndexedSeq(0))
+                    .map(x => Right(Some(x)))
+                    .getOrElse(Left(Exporter.InvalidRowId))
+          case None => // no customer row identifier.  Use system row identifier.
+            val rowIdRep = common.jsonReps(ctx.copyCtx.datasetInfo)(ctx.copyCtx.pkCol_!.typ)
+            rowIdRep.fromJValue(JString(rid)).map(x => Right(Some(x))).getOrElse(Left(Exporter.InvalidRowId))
         }
       }
+    rowIdPossiblyError match {
```

Comment:
Could you not return Right or Left above in the Match cases or will it make things overly complicated?